advanced-topic (#5)
vaibhav-yb/yugabyte-db: docs/content/preview/explore/change-data-capture/using-logical-replication/advanced-topic.md

    identifier: advanced-topics
    weight: 50
type: docs
---

In this section, we explore a range of topics that are designed to provide deeper insights and enhance your understanding of advanced functionalities. 

> ## Schema Evolution

Reviewer: ```suggestion
## Schema evolution
```

> A change in the schema of the tables (Alter Table) being streamed are transparently handled by the database without manual intervention.

Reviewer: ```suggestion
A change in the schema of the tables (Alter Table) being streamed is transparently handled by the database without manual intervention.
```

> This is illustrated in the below example. Note that the client being used for example purposes is `pg_recvlogical` which is discussed in [this section](../using-logical-replication/get-started.md).

Reviewer: ```suggestion
This is illustrated in the below example. Note that the client being used for example purposes is `pg_recvlogical` which is discussed in [this section](../using-logical-replication/get-started).
```

```sh
# Create table and create the replication slot. pg_recvlogical uses test_decoding output plugin by default.
CREATE TABLE demo_table (id INT PRIMARY KEY);

pg_recvlogical -d yugabyte --slot=demo_slot --create-slot

# Start streaming the changes from the replication slot.
pg_recvlogical -d yugabyte --slot=demo_slot --start -f -

# In a new shell, start ysqlsh and insert some data.
bin/ysqlsh

INSERT INTO demo_table VALUES (1);
INSERT INTO demo_table VALUES (2);

# pg_recvlogical would receive the inserts and print it on the console.
BEGIN 2
TABLE public.demo_table: INSERT: id[integer]:1
COMMIT 2
BEGIN 3
TABLE public.demo_table: INSERT: id[integer]:2
COMMIT 3

# Add a new column to the demo_table and insert some more rows.
ALTER TABLE demo_table ADD COLUMN address TEXT;
INSERT INTO demo_table VALUES (3, 'address1');
INSERT INTO demo_table VALUES (4, 'address2');

# Without any manual intervention pg_recvlogical would receive the inserts with the new schema and print it on the console.
BEGIN 4
TABLE public.demo_table: INSERT: id[integer]:3 col_text[text]:'address1'
COMMIT 4
BEGIN 5
TABLE public.demo_table: INSERT: id[integer]:4 col_text[text]:'address2'
COMMIT 5

> ```

Reviewer: I think these sections should be separated, we should not mix SQL commands with shell commands.

Reviewer: Marking this as resolved since the work will be taken up as part of another PR as discussed offline.

> ## Addition of Tables to Publication

Reviewer: ```suggestion
## Addition of tables to publication
```

> Addition of tables to the streaming list after slot creation is currently a preview feature. Inorder to enable dynamic table addition the tserver preview flag [cdcsdk_enable_dynamic_table_support](../../../../reference/configuration/yb-tserver/#cdcsdk_enable_dynamic_table_support) should be set to true.

Reviewer: ```suggestion
Addition of tables to the streaming list after slot creation is currently a preview feature. In order to enable dynamic table addition the tserver preview flag [cdcsdk_enable_dynamic_table_support](../../../../reference/configuration/yb-tserver/#cdcsdk_enable_dynamic_table_support) should be set to true.
```

The Publication’s tables list can change in two cases. The first case is when a previously created table is added to the publication by performing an alter publication. 

```sql
CREATE TABLE test_table_1(id INT PRIMARY KEY, aa INT, bb INT);
CREATE TABLE test_table_2(id INT PRIMARY KEY, aa INT, bb INT);

CREATE PUBLICATION PUB FOR TABLE test_table_1;

-- Start consumption through a replication slot.

ALTER PUBLICATION ADD TABLE test_table_2;

> 

Reviewer: ```suggestion
```

```

> The second case is when a dynamically created table is added to ALL TABLES publication upon creation.

Reviewer: ```suggestion
The second case is when a dynamically created table is added to `ALL TABLES` publication upon creation.
```

```sql
CREATE TABLE test_table_1(id INT PRIMARY KEY, aa INT, bb INT);

CREATE PUBLICATION PUB FOR ALL TABLES;

-- Start consumption through a replication slot.

CREATE TABLE test_table_2(id INT PRIMARY KEY, aa INT, bb INT);

```

> 

Reviewer: ```suggestion
```

> Unlike Postgres, any changes made to the publication’s tables list will not be applied immediately in YB. Instead the publication’s tables list will be periodically refreshed and changes, if any, will be applied. The refresh interval is denoted by the flag [cdcsdk_publication_list_refresh_interval_secs](../../../../reference/configuration/yb-tserver/#cdcsdk_publication_list_refresh_interval_secs) and has a default value of one hour (3600 sec). This means that any changes made to the publication’s tables list will be applied after `cdcsdk_publication_list_refresh_interval_secs` in the worst case. Consider the following example:

Reviewer: ```suggestion
Unlike Postgres, any changes made to the publication’s tables list will not be applied immediately in YugabyteDB. Instead the publication’s tables list will be periodically refreshed and changes, if any, will be applied. The refresh interval is denoted by the flag [cdcsdk_publication_list_refresh_interval_secs](../../../../reference/configuration/yb-tserver/#cdcsdk_publication_list_refresh_interval_secs) and has a default value of one hour (3600 sec). This means that any changes made to the publication’s tables list will be applied after `cdcsdk_publication_list_refresh_interval_secs` in the worst case. Consider the following example:
```
